### ISSUE TITLE
Three things

### ISSUE THREAD

**Issue by philipturner** [vojtamolda/Sundials.swift#2]

**Comment by vojtamolda** [comment]
Hello @philipturner,

Thanks a lot for posting the message here! I'm glad that S4TF is being worked on again. As far as I know, work on the differentiation in the compiler itself actually never stopped. A lot of S4TF team members found a way to continue elsewhere. So to be honest, I'm not sure if I support the resurrection of the whole TensorFlow project. The XLA part of is a pretty scary aberration. Don't get me wrong, though, I'll be very happy if it happens, but I won't be able to invest any effort.

I personally think that Swift needs a multi-dimensional array type, aka the only the `Tensor` struct from TensorFlow. Ideally with hardware acceleration support where possible. It would serve the same role as `numpy.ndarray` in the Python ecosystem. Numerical libraries (like this one) could be built on top of it and would become compatible with each other. Right now people have to roll their own half-baked implementations and the situation is far from ideal.

1) Nice fix for the meantime!

2) Sorry for the broken link. The whole post needs to be updated anyway. I'll get to it eventually...

3) I don't have a problem with unarchiving the repo. Hope it helps at least little bit.

**Comment by philipturner** [comment]
I'm not asking for your assistance fixing the code of S4TF. I'm just asking that you contribute by unarchiving the repository. The S4TF team and I are working on autodiff and XLA!

**Comment by philipturner** [comment]
Linking this thread to https://github.com/tensorflow/swift/issues/623 to put further momentum behind the effort to resurrect S4TF

**Comment by vojtamolda** [comment]
I didn't mean the `ShapeArray`; I view it's purpose only as an escape hatch to bridge different `Tensor`-ish type implementations and convert data back and forth. I was talking about a single generally used and versatile implementation like the `ndarray` from Python.

The [tensor-thumbnail](https://github.com/vojtamolda/tensor-thumbnail) repo was already unarchived when I posted my first comment. So I think we're done here.

**Comment by philipturner** [comment]
Thanks for unarchiving!

**Comment by RahulBhalley** [comment]
@philipturner I doubt if Google will agree to unarchive the S4TF project. Have you had any conversations with the team upon unarchiving S4TF approval? I believe they're busy doing their own new jobs in Google and won't be able to work on this project anymore like handling PRs/issues/no design meetings and more. Also it's **very very much possible (almost 100%)** that the root-level access to [swift-apis](https://github.com/tensorflow/swift-apis) or any other repository will not be given.

> I was talking about a single generally used and versatile implementation like the `ndarray` from Python.

I believe we should instead create a new differentiable `Tensor` library (with Metal support and later CUDA / XLA (that can be also adopted from the current state of S4TF but I think multi-TPU cores support was not made available)) using Swift development toolchain and distribute it as a package. Then build a deep learning library on top of it. This way, in future, the differentiable `Tensor` library can be used for building other packages too.

**Comment by philipturner** [comment]
I have had plenty of conversations with the S4TF team and I know what I'm doing. Some of them are thrilled that I'm putting effort into S4TF again, and I see the resurrection as a realistic goal. The reason I'm stirring up so much momentum from the open-source community is to counter a potential resistance from Google.

**Comment by philipturner** [comment]
@vojtamolda I'm putting a hold on the Swift Forums post, given that both you and @RahulBhalley seem to have mixed feelings about the topic.

**Comment by vojtamolda** [comment]
Please, don't let a grumpy dude like me slow you down ;) I'm honestly really happy that you're doing what you're doing. It's just that I'm not big on social media or discussion forums. I don't feel like posting about un-archiving of a tiny library that  has 0 users besides myself.

I agree with all @RahulBhalley wrote. The most useful project is to rip the `Tensor` type out of TensorFlow and revive it with a new Metal backend. And I don't think you need to involve Google in this.

The unique part about S4TF is differentiation of arbitrary code and structs. To me it allowed differentiable simulations which is something that was very hard to do before. If one wants to use S4TF just for machine learning there is just too many obstacles and sharp edges that in the end don't pay off. People are better off staying in the Python world for pure ML and it doesn't make sense to sail against this wind.

However, if you want to do specialized things, the differentiability built into the compiler is very unique. There's not that many other projects that allow you to do it. 
So to me it makes sense to leverage this niche of differentiable numerical algorithms that isn't possible to replicate in JAX/PyTorch/TensorFlow. I believe this is where the next generation of numerical solvers gets created.

**Comment by philipturner** [comment]
> People are better off staying in the Python world for pure ML and it doesn't make sense to sail against this wind.

I disagree with that statement. I think there's a good chance I'll beat Torch's new Metal backend in performance, because I have much more experience with Metal than the people currently working on it. Also, you don't need to go away from Python. I just revived support for Swift in Google Colab (postponing official release of this until SPM support is reworked). I did a heck ton of bridging between Swift and Python.

I also realized that CoreMLTools can be used entirely from Swift because it doesn't require subclassing or lambdas. Rather, Swift has a unique advantage because of its seamless integration with Python and unique differentiation capabilities. I even once contemplated using Python TensorFlow from Swift!

**Comment by philipturner** [comment]
> One alternative idea to reviving the Tensor type I have is to use the ArrayFire library, wrap it in Swift and then write custom backward passes for all ops.

Oh no... @palle-k and @compnerd have advised me otherwise. It's in C++ and has terrible performance, which is a non-starter. Rather, I'm looking for some common interface (XLA) where Metal and DirectML can both interface with it. DirectML would actually be easy (although only as eager mode because it lacks something like MPSGraph) to integrate as it's more feature complete than MPS. However, I'm going for Metal for the time being.

**Comment by philipturner** [comment]
>  I don't feel like posting about un-archiving of a tiny library that has 0 users besides myself.

Your library has more users than my [differentiation](https://github.com/philipturner/differentiation) or [swift-colab](https://github.com/philipturner/swift-colab). And this is actually a really big milestone in the resurrection of S4TF. I was able to convince someone with a library dedicated to S4TF to go along with this.

I can make the post for you and planned to do that. It will be short and isn't social media. Trust me, there's nothing to worry about.

**Comment by philipturner** [comment]
Torch’s current Metal support is for inference, not training. That’s comparably easy to accomplish and has been done countless times. The real problem is using Metal for training.

**Comment by philipturner** [comment]
To be honest, I’m not really that interested in machine learning or differentiation. In November, I set out with one goal in mind: resurrect Swift for TensorFlow, for these of resurrecting Swift for TensorFlow.

**Comment by RahulBhalley** [comment]
> Also, their Metal code unnecessarily uses Objective-C.

I wonder what would've been a better approach instead of Objective-C++ when interfacing C++ Tensor library ATen to Metal Performance Shaders?

**Comment by philipturner** [comment]
1) C is the alternative. Bridge through Core Foundation, and you never have to touch ObjC. I did tons of C interfacing to bridge between Swift and Python.

2) There’s more than just calling backward() to do training. You have to entirely rethink algorithms such as matrix multiplication (SGEMV becomes SGEMM) because it’s batches. Also, you can heavily optimize convolutions by using batches im2col.

**Comment by philipturner** [comment]
@RahulBhalley did you read my messages on the MetalXLA thread? If you aren’t interested in helping out, I’d like to know sooner rather than later for planning purposes.
